### PR TITLE
[DependencyInjection] Handle env var placeholders in CheckTypeDeclarationsPass

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

A case we forgot to handle.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php [hunk 1]
```diff
@@ -172,8 +175,12 @@ private function checkType(Definition $checkedDefinition, $value, \ReflectionPar
 
         if ($value instanceof Parameter) {
             $value = $this->container->getParameter($value);
-        } elseif (\is_string($value) && '%' === ($value[0] ?? '') && preg_match('/^%([^%]+)%$/', $value, $match)) {
-            $value = $this->container->getParameter($match[1]);
+        } elseif (\is_string($value)) {
+            if ('%' === ($value[0] ?? '') && preg_match('/^%([^%]+)%$/', $value, $match)) {
+                $value = $this->container->getParameter($match[1]);
+            } elseif ($envPlaceholderUniquePrefix && 0 === strpos($value, $envPlaceholderUniquePrefix)) {
+                $value = $this->container->resolveEnvPlaceholders($value, true);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php]: this means the command will fail if an env var is missing - not what I would expect from a linter
what about skipping these instead?

Comment (fancyweb) [reply src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php]: Yes that will fail and that's fine to me. A missing env var is a problem to validate the container. Maybe we need a strict and lax mode.

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php]: That's not fine to me: linting should be possible in build environments, that don't have env vars available.

Comment (fancyweb) [reply src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php]: What about skipping them by default, but add a "--strict" option then? Resolving env vars is a legitimate case.

**Comment by nicolas-grekas** [comment]
There's something we do in Config: we replace env placeholders with credible values.
But honestly, it can only go back to false-positives. And the false-negatives don't matter: as long as the app works, it doesn't exist.

The goal of the command is not to prove a theoretical point but to help spot errors quickly.

We should skip instead.

File: src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php [hunk 2]
```diff
@@ -174,12 +178,24 @@ private function checkType(Definition $checkedDefinition, $value, \ReflectionPar
             throw new InvalidParameterTypeException($this->currentId, $class, $parameter);
         }
 
-        if ($value instanceof Parameter) {
```
Comment (fancyweb) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php]: Loaders never instantiate `Parameter` so we don't need it.

File: src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php [hunk 3]
```diff
@@ -174,12 +178,24 @@ private function checkType(Definition $checkedDefinition, $value, \ReflectionPar
...
+        } elseif (\is_string($value)) {
+            if ('%' === ($value[0] ?? '') && preg_match('/^%([^%]+)%$/', $value, $match)) {
+                // Only array parameters are not inlined when dumped.
+                $value = [];
+            } elseif ($envPlaceholderUniquePrefix && false !== strpos($value, 'env_')) {
+                // If the value is an env placeholder that is either mixed with a string or with another env placeholder, then its resolved value will always be a string, so we don't need to resolve it.
+                // We don't need to change the value because it is already a string.
+                if ('' === preg_replace('/'.$envPlaceholderUniquePrefix.'_\w+_[a-f0-9]{32}/U', '', $value, -1, $c) && 1 === $c) {
+                    try {
+                        $value = $this->container->resolveEnvPlaceholders($value, true);
```
Comment (fancyweb) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php]: @nicolas-grekas Let's try to resolve the env an skip it if it fails?

File: src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php [hunk 4]
```diff
@@ -174,12 +178,24 @@ private function checkType(Definition $checkedDefinition, $value, \ReflectionPar
...
-        if ($value instanceof Parameter) {
-            $value = $this->container->getParameter($value);
-        } elseif ($value instanceof Expression) {
+        if ($value instanceof Expression) {
             $value = $this->getExpressionLanguage()->evaluate($value, ['container' => $this->container]);
-        } elseif (\is_string($value) && '%' === ($value[0] ?? '') && preg_match('/^%([^%]+)%$/', $value, $match)) {
-            $value = $this->container->getParameter($match[1]);
+        } elseif (\is_string($value)) {
+            if ('%' === ($value[0] ?? '') && preg_match('/^%([^%]+)%$/', $value, $match)) {
+                // Only array parameters are not inlined when dumped.
```
Comment (fancyweb) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php]: Added comments because those things are kind of hard to follow. We actually don't need to resolve the value here since we only care about the type.

File: src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php [hunk 5]
```diff
@@ -174,12 +178,24 @@ private function checkType(Definition $checkedDefinition, $value, \ReflectionPar
...
+                $value = [];
+            } elseif ($envPlaceholderUniquePrefix && false !== strpos($value, 'env_')) {
+                // If the value is an env placeholder that is either mixed with a string or with another env placeholder, then its resolved value will always be a string, so we don't need to resolve it.
+                // We don't need to change the value because it is already a string.
+                if ('' === preg_replace('/'.$envPlaceholderUniquePrefix.'_\w+_[a-f0-9]{32}/U', '', $value, -1, $c) && 1 === $c) {
```
Comment (fancyweb) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php]: 3 cases here (with `param: "%env(FOO)%"`:
- only 1 env placeholder (eg: `%param%`) -> we try to resolve it to get its type.
- more than 1 chained env placeholder (eg: `%param%%param%`) -> it has to be a string
- mixed env placeholder with a string (eg : `foo%param%`) -> it has to be a string

File: src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php]: let's try

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @fancyweb.
